### PR TITLE
Generate cross-schema relations in print schema

### LINKED ISSUE #4476: diesel cli generate does not create relations for models in different schema when using postgres schema
### Versions

- **Rust:** 1.84.1
- **Diesel:** 2.2.7
- **Database:** Postgres 17.1
- **Operating System** Ubuntu 24.04.1

### Feature Flags

- **diesel:** ["postgres_backend", "postgres", "time", "r2d2", "uuid", "serde_json"]

## Problem Description
Similar to https://github.com/diesel-rs/diesel/issues/2096 
I am however using the schema="name" found in the resolution of that issue as follows 

```toml
[print_schema.game]
file = "src/diesel_schema/game.rs"
custom_type_derives = ["diesel::query_builder::QueryId", "Clone"]
schema = "game"

[print_schema.people]
file = "src/diesel_schema/people.rs"
custom_type_derives = ["diesel::query_builder::QueryId", "Clone"]
schema = "people"

[migrations_directory]
dir = "./migrations"
```
I also have 2 tables, one in each schema:
```sql
CREATE TABLE "people"."api_token"
(
    id       INTEGER PRIMARY KEY GENERATED ALWAYS AS IDENTITY,
    val      UUID NOT NULL DEFAULT gen_random_uuid()
);
```

```sql 
CREATE TABLE "game"."game_session"
(
    id INTEGER PRIMARY KEY GENERATED ALWAYS AS IDENTITY,
    api_token_id INTEGER REFERENCES "people"."api_token" (id) NOT NULL -- relation
);
```

The generated diesel schemas do not include the relationship:
```rust
pub mod game {
    diesel::table! {
        game.game_session (id) {
            id -> Int4,
            api_token_id -> Int4,
        }
    }
}
```
```rust 
pub mod people {
    diesel::table! {
        people.api_token (id) {
            id -> Int4,
            val -> Uuid,
        }
    }
}
```

### What is the expected output?

one of the schema(game.rs I think?) should include:
```rust
diesel::joinable!(game_session -> api_token (api_token_id));
diesel::allow_tables_to_appear_in_same_query!(
        api_token,
        game_session
}
```

### Are you seeing any additional errors?
Yes, cli `generate --diff-schema` produces empty sql files when I manually add the above lines and then call it.

### Steps to reproduce
see above

## Checklist

- [x] I have already looked over the [issue tracker](https://github.com/diesel-rs/diesel/issues) and the [discussion forum](https://github.com/diesel-rs/diesel/discussions) for similar possible closed issues.
 
- [x] This issue can be reproduced on Rust's stable channel. (Your issue will be
  closed if this is not the case)
- [x] This issue can be reproduced without requiring a third party crate

### PR DESCRIPTION
**Pull request by henyanagar**
Fixes #4476
I looked into the schema generation flow for multiple **print_schema** sections and found that cross-schema foreign keys were filtered out before Diesel generated the relation macros.

The issue happens when one generated schema file contains the child table and another generated schema file contains the referenced table. In that case, the foreign key exists in PostgreSQL, but the generated Diesel schema did not include the expected **joinable!** and **allow_tables_to_appear_in_same_query**! entries.

**Changes**
This PR changes the foreign key loading/filtering logic so that relations between generated schemas are kept, while still only emitting **joinable!** in the schema file that contains the child table.
I also added a regression test with two PostgreSQL schemas:

- game.game_session
- people.api_token

where **game.game_session.api_token_id** references **people.api_token.id**.

**AI/LLM disclosure:**
I used an LLM(Claude) as a helper while investigating the code path and checking the change. I reviewed the suggested changes, made sure I understood them, and ran the tests locally.

**Tests I ran:**
-  cargo test -p diesel_cli --no-default-features --features postgres --test tests print_schema_with_multiple_schema_cross_schema_foreign_key -- --nocapture
-  cargo test -p diesel_cli --no-default-features --features postgres --test tests print_schema_with_multiple_schema -- --nocapture
- cargo test -p diesel_cli --no-default-features --features postgres --test tests print_schema_with_foreign_keys -- --nocapture
- cargo test -p diesel_cli --no-default-features --features postgres --test tests print_schema_specifying_schema_name_with_foreign_keys -- --nocapture

### CODE REVIEW

File: diesel_cli/tests/print_schema.rs [hunk 1]
```diff
@@ -450,6 +450,34 @@ fn print_schema_with_multiple_schema() {
...
+    let schema = p.file_contents("src/schema1.rs").replace("\r\n", "\n");
+    assert!(
+        schema.contains("diesel::joinable!(game_session -> api_token (api_token_id));"),
+        "expected schema1.rs to contain a joinable! for the cross-schema foreign key, got:\n{schema}"
+    );
+    assert!(
+        schema.contains("diesel::allow_tables_to_appear_in_same_query!(api_token, game_session,);"),
+        "expected schema1.rs to contain allow_tables_to_appear_in_same_query! for the cross-schema foreign key, got:\n{schema}"
+    );
+}
```
Comment (weiznich) [inline diesel_cli/tests/print_schema.rs:494]: This test should use the existing `test_multiple_print_schema()` function and the existing dictory structure instead of reimplementing the same logic here again slightly different.

File: diesel_cli/src/print_schema.rs [hunk 2]
```diff
@@ -513,13 +554,17 @@ pub fn output_schema(
     );
 
     let foreign_keys = load_foreign_key_constraints(connection, config.schema_name())?;
-    let safe_tables =
-        &filter_column_structure(&table_names, SupportedQueryRelationStructures::Table);
+    let current_schema_safe_tables =
+        filter_column_structure(&table_names, SupportedQueryRelationStructures::Table);
+    let fk_safe_tables = multi_schema_safe_tables.unwrap_or(&current_schema_safe_tables);
```
Comment (weiznich) [inline diesel_cli/src/print_schema.rs]: Use `unwrap_or_else(|| …)` here to only load the tables if `multi_schema_safe_tables` is not provided. That might require a bit additional wrapping to get the result and borrowing right but that should be doable by just wrapping the `multi_schema_safe_tables` into a result via `.map(Ok)` and also by using `Cow`.

**Review by weiznich — CHANGES_REQUESTED** [review]
Thanks for opening this PR. In general I'm open to add this feature, but there are a few minor things in the PR that I would like to see adjusted.

It also needs to have a change log entry, so that users are aware of this change.

**Comment by henyanagar** [comment]
Thanks for the review.

I updated the PR according to your comments. The test now uses the existing test_multiple_print_schema  helper with snapshots, I added a changelog entry, and I changed the  unwrap_or  part to use  Cow  / unwrap_or_else .
I also ran the related PostgreSQL print-schema tests locally, and they passed.

**Review by weiznich — CHANGES_REQUESTED** [review]
I tested this locally and unfortunately the generated schema doesn't compile with the following error messages:

```
error[E0433]: failed to resolve: use of unresolved module or unlinked crate `api_token`
  --> src/schema1.rs:24:39
   |
24 |     diesel::joinable!(game_session -> api_token (api_token_id));
   |                                       ^^^^^^^^^ use of unresolved module or unlinked crate `api_token`
   |
   = help: if you wanted to use a crate named `api_token`, use `cargo add api_token` to add it to your `Cargo.toml`
help: consider importing this struct through its public re-export
   |
 4 +     use crate::schema2::people::api_token::dsl::api_token;
   |

error[E0433]: failed to resolve: use of unresolved module or unlinked crate `api_token`
  --> src/schema1.rs:26:51
   |
26 |     diesel::allow_tables_to_appear_in_same_query!(api_token, game_session,);
   |                                                   ^^^^^^^^^ use of unresolved module or unlinked crate `api_token`
   |
   = help: if you wanted to use a crate named `api_token`, use `cargo add api_token` to add it to your `Cargo.toml`
help: consider importing this struct through its public re-export
   |
 4 +     use crate::schema2::people::api_token::dsl::api_token;
   |

error[E0433]: failed to resolve: use of unresolved module or unlinked crate `game_session`
  --> src/schema2.rs:18:62
   |
18 |     diesel::allow_tables_to_appear_in_same_query!(api_token, game_session,);
   |                                                              ^^^^^^^^^^^^ use of unresolved module or unlinked crate `game_session`
   |
   = help: if you wanted to use a crate named `game_session`, use `cargo add game_session` to add it to your `Cargo.toml`
help: consider importing this struct through its public re-export
   |
 4 +     use crate::schema1::game::game_session::dsl::game_session;
   |

error[E0119]: conflicting implementations of trait `AppearsInFromClause<game_session::table>` for type `game_session::table`
  --> src/schema1.rs:4:5
   |
 4 | /     diesel::table! {
 5 | |         /// Representation of the `game.game_session` table.
 6 | |         ///
 7 | |         /// (Automatically generated by Diesel.)
...  |
22 | |     }
   | |_____^
   |
   = note: conflicting implementation in crate `diesel`:
           - impl<T1, T2> AppearsInFromClause<T2> for T1
             where T1: TableNotEqual<T2>, T1: Table, T2: Table;
   = note: this error originates in the macro `diesel::table` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0119]: conflicting implementations of trait `AppearsInFromClause<Alias<_>>` for type `game_session::table`
  --> src/schema1.rs:4:5
   |
 4 | /     diesel::table! {
 5 | |         /// Representation of the `game.game_session` table.
 6 | |         ///
 7 | |         /// (Automatically generated by Diesel.)
...  |
22 | |     }
   | |_____^
   |
   = note: conflicting implementation in crate `diesel`:
           - impl<T1, T2> AppearsInFromClause<T2> for T1
             where T1: TableNotEqual<T2>, T1: Table, T2: Table;
   = note: upstream crates may add a new impl of trait `diesel::Table` for type `diesel::query_source::Alias<_>` in future versions
   = note: this error originates in the macro `diesel::table` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0119]: conflicting implementations of trait `AppearsInFromClause<api_token::table>` for type `api_token::table`
  --> src/schema2.rs:4:5
   |
 4 | /     diesel::table! {
 5 | |         /// Representation of the `people.api_token` table.
 6 | |         ///
 7 | |         /// (Automatically generated by Diesel.)
...  |
16 | |     }
   | |_____^
   |
   = note: conflicting implementation in crate `diesel`:
           - impl<T1, T2> AppearsInFromClause<T2> for T1
             where T1: TableNotEqual<T2>, T1: Table, T2: Table;
   = note: this error originates in the macro `diesel::table` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0119]: conflicting implementations of trait `AppearsInFromClause<Alias<_>>` for type `api_token::table`
  --> src/schema2.rs:4:5
   |
 4 | /     diesel::table! {
 5 | |         /// Representation of the `people.api_token` table.
 6 | |         ///
 7 | |         /// (Automatically generated by Diesel.)
...  |
16 | |     }
   | |_____^
   |
   = note: conflicting implementation in crate `diesel`:
           - impl<T1, T2> AppearsInFromClause<T2> for T1
             where T1: TableNotEqual<T2>, T1: Table, T2: Table;
   = note: upstream crates may add a new impl of trait `diesel::Table` for type `diesel::query_source::Alias<_>` in future versions
   = note: this error originates in the macro `diesel::table` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0119]: conflicting implementations of trait `AliasAppearsInFromClause<_, game_session::table>` for type `game_session::table`
  --> src/schema1.rs:4:5
   |
 4 | /     diesel::table! {
 5 | |         /// Representation of the `game.game_session` table.
 6 | |         ///
 7 | |         /// (Automatically generated by Diesel.)
...  |
22 | |     }
   | |_____^
   |
   = note: conflicting implementation in crate `diesel`:
           - impl<T1, T2, S> AliasAppearsInFromClause<S, T2> for T1
             where <S as AliasSource>::Target == T1, T1: TableNotEqual<T2>, T1: Table, T2: Table, S: AliasSource;
   = note: this error originates in the macro `diesel::table` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0119]: conflicting implementations of trait `AliasAppearsInFromClause<_, api_token::table>` for type `api_token::table`
  --> src/schema2.rs:4:5
   |
 4 | /     diesel::table! {
 5 | |         /// Representation of the `people.api_token` table.
 6 | |         ///
 7 | |         /// (Automatically generated by Diesel.)
...  |
16 | |     }
   | |_____^
   |
   = note: conflicting implementation in crate `diesel`:
           - impl<T1, T2, S> AliasAppearsInFromClause<S, T2> for T1
             where <S as AliasSource>::Target == T1, T1: TableNotEqual<T2>, T1: Table, T2: Table, S: AliasSource;
   = note: this error originates in the macro `diesel::table` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0119]: conflicting implementations of trait `AliasAliasAppearsInFromClause<game_session::table, _, _>` for type `game_session::table`
  --> src/schema1.rs:4:5
   |
 4 | /     diesel::table! {
 5 | |         /// Representation of the `game.game_session` table.
 6 | |         ///
 7 | |         /// (Automatically generated by Diesel.)
...  |
22 | |     }
   | |_____^
   |
   = note: conflicting implementation in crate `diesel`:
           - impl<T1, T2, S1, S2> AliasAliasAppearsInFromClause<T1, S2, S1> for T2
             where <S1 as AliasSource>::Target == T1, <S2 as AliasSource>::Target == T2, T1: TableNotEqual<T2>, T1: Table, T2: Table, S1: AliasSource, S2: AliasSource;
   = note: this error originates in the macro `diesel::table` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0119]: conflicting implementations of trait `AliasAliasAppearsInFromClause<api_token::table, _, _>` for type `api_token::table`
  --> src/schema2.rs:4:5
   |
 4 | /     diesel::table! {
 5 | |         /// Representation of the `people.api_token` table.
 6 | |         ///
 7 | |         /// (Automatically generated by Diesel.)
...  |
16 | |     }
   | |_____^
   |
   = note: conflicting implementation in crate `diesel`:
           - impl<T1, T2, S1, S2> AliasAliasAppearsInFromClause<T1, S2, S1> for T2
             where <S1 as AliasSource>::Target == T1, <S2 as AliasSource>::Target == T2, T1: TableNotEqual<T2>, T1: Table, T2: Table, S1: AliasSource, S2: AliasSource;
   = note: this error originates in the macro `diesel::table` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0119]: conflicting implementations of trait `FieldAliasMapperAssociatedTypesDisjointnessTrick<game_session::table, _, _>` for type `game_session::table`
  --> src/schema1.rs:4:5
   |
 4 | /     diesel::table! {
 5 | |         /// Representation of the `game.game_session` table.
 6 | |         ///
 7 | |         /// (Automatically generated by Diesel.)
...  |
22 | |     }
   | |_____^
   |
   = note: conflicting implementation in crate `diesel`:
           - impl<TS, TC, S, C> FieldAliasMapperAssociatedTypesDisjointnessTrick<TC, S, C> for TS
             where <S as AliasSource>::Target == TS, <C as Column>::Table == TC, S: AliasSource, C: Column, TC: Table, TS: TableNotEqual<TC>;
   = note: this error originates in the macro `diesel::table` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0119]: conflicting implementations of trait `FieldAliasMapperAssociatedTypesDisjointnessTrick<api_token::table, _, _>` for type `api_token::table`
  --> src/schema2.rs:4:5
   |
 4 | /     diesel::table! {
 5 | |         /// Representation of the `people.api_token` table.
 6 | |         ///
 7 | |         /// (Automatically generated by Diesel.)
...  |
16 | |     }
   | |_____^
   |
   = note: conflicting implementation in crate `diesel`:
           - impl<TS, TC, S, C> FieldAliasMapperAssociatedTypesDisjointnessTrick<TC, S, C> for TS
             where <S as AliasSource>::Target == TS, <C as Column>::Table == TC, S: AliasSource, C: Column, TC: Table, TS: TableNotEqual<TC>;
   = note: this error originates in the macro `diesel::table` (in Nightly builds, run with -Z macro-backtrace for more info)

Some errors have detailed explanations: E0119, E0433.
For more information about an error, try `rustc --explain E0119`.
error: could not compile `testing` (bin "testing") due to 13 previous errors
```

The main problems are: 
* The imported tables from the other schemas are not in scope. (References `api_token`, but that's not imported)
* It generates `diesel::allow_tables_to_appear_in_same_query!(api_token, game_session,)` in both schema files, resulting in conflicting implementation of several traits.

Both issues need to be fixed before merging this PR.

File: diesel_cli/src/print_schema.rs [hunk 3]
```diff
@@ -966,10 +1101,42 @@ impl<'a> Display for QueryRelationDefinitions<'a> {
...
             }
             AllowTablesToAppearInSameQueryConfig::None => vec![],
         };
+        let table_groups = if self.multi_schema_module_prefixes.is_some() {
+            table_groups
+                .into_iter()
+                .filter(|table_group| {
+                    let all_local = table_group
+                        .iter()
+                        .all(|table| table.schema.as_deref() == self.current_pg_schema);
```
Comment (weiznich) [inline diesel_cli/src/print_schema.rs]: This is not necessary correct. Two print-schema "subschemas" could contain tables from the same database schema, only with different set of filters. This would need a better more robust filter setup.

File: diesel_cli/src/print_schema.rs [hunk 4]
```diff
@@ -499,10 +523,89 @@ fn escape_rust_string(input: &str) -> String {
...
+}
+
+fn table_codegen_path<'a>(
+    table: &'a TableName,
+    current_pg_schema: Option<&'a str>,
+    module_prefixes: Option<&'a BTreeMap<String, String>>,
+    use_file_module_paths: bool,
+) -> Cow<'a, str> {
+    match (current_pg_schema, module_prefixes) {
+        (Some(current), Some(prefixes)) if table.schema.as_deref() != Some(current) => {
```
Comment (weiznich) [inline diesel_cli/src/print_schema.rs]: As with the table group filtering: A rust schema file could contain tables from the same database schema, just with different sets of include/exclude, so this is not the correct way to filter the right tables for this file.

This case should also be covered with an additional test.

**Review by weiznich — CHANGES_REQUESTED** [review]
Thanks for the update. This looks better now, but the filtering for which table belongs to which schema.rs file is still not covering all edge cases. This needs to be addressed + covered with tests before merging.

**Review by weiznich — APPROVED** [review]
Thanks for the update.

I pushed a minimal change to explain why the project_name truncation is required and made it also a bit more robust for future even longer names.
